### PR TITLE
[Modules] New module for `Microsoft.Search/searchServices`

### LINKED ISSUE #2183: [Modules] Create new `Microsoft.Search/searchServices` module
Create a new CARML module for the `Microsoft.Search/searchServices` resource type as per the [service documentation](https://learn.microsoft.com/en-us/azure/templates/microsoft.search/searchservices?pivots=deployment-language-bicep) and our [contribution guide](https://github.com/Azure/ResourceModules/wiki/Contribution%20guide).

### PR DESCRIPTION
**Pull request by krbar**
# Description

This PR resolves #2183 and adds a new CARML module for `Microsoft.Search/searchServices`.

## Pipeline references

| Pipeline |
| - |
| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=138&branchName=users%2Fkrbar%2F2183-searchservices-module) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/search/search-service/.test/common/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.search.searchservice-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'sesvcom'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/common/main.test.bicep:15]:
```suggestion
param serviceShort string = 'ssscom'
```

File: modules/search/search-service/.test/min/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,45 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.search.searchservice-${serviceShort}-rg'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/min/main.test.bicep:9]:
```suggestion
param resourceGroupName string = 'ms.search.searchservices-${serviceShort}-rg'
```

File: modules/search/search-service/.test/common/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,110 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.search.searchservice-${serviceShort}-rg'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/common/main.test.bicep:9]:
```suggestion
param resourceGroupName string = 'ms.search.searchservices-${serviceShort}-rg'
```

File: modules/search/search-service/.test/min/main.test.bicep [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'sesvmin'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/min/main.test.bicep:15]:
```suggestion
param serviceShort string = 'sssmin'
```

File: modules/search/search-service/.test/pe/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,95 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.search.searchservice-${serviceShort}-rg'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/pe/main.test.bicep:9]:
```suggestion
param resourceGroupName string = 'ms.search.searchservices-${serviceShort}-rg'
```

File: modules/search/search-service/.test/pe/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'sesvpe'
```
Comment (ahmadabdalla) [inline modules/search/search-service/.test/pe/main.test.bicep:15]:
```suggestion
param serviceShort string = 'ssspe'
```

File: modules/search/search-service/main.bicep [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+//   Parameters   //
+// ============== //
+
+@description('Required. The name of the Azure Cognitive Search service to create or update. Search service names must only contain lowercase letters, digits or dashes, cannot use dash as the first two or last one characters, cannot contain consecutive dashes, and must be between 2 and 60 characters in length. Search service names must be globally unique since they are part of the service URI (https://<name>.search.windows.net). You cannot change the service name after the service is created.')
+param name string
+
+@description('Optional. Defines the options for how the data plane API of a Search service authenticates requests. Must remain an empty object {} if \'disableLocalAuth\' is set to true.')
+param authOptions object = {}
+
+@description('Optional. The name of logs that will be streamed.')
```
Comment (ahmadabdalla) [inline modules/search/search-service/main.bicep:15]: suggest moving the diagnostics parameters lower in this list

File: modules/search/search-service/main.bicep [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+@description('Optional. Enable telemetry via the Customer Usage Attribution ID (GUID).')
+param enableDefaultTelemetry bool = true
+
+@description('Optional. Describes a policy that determines how resources within the search service are to be encrypted with Customer Managed Keys.')
+@allowed([
+  'Disabled'
+  'Enabled'
+  'Unspecified'
+])
+param cmkEnforcement string = 'Unspecified'
```
Comment (ahmadabdalla) [inline modules/search/search-service/main.bicep:58]: can we test the `Enabled` use case?

File: modules/search/search-service/main.bicep [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+@description('Optional. Applicable only for the standard3 SKU. You can set this property to enable up to 3 high density partitions that allow up to 1000 indexes, which is much higher than the maximum indexes allowed for any other SKU. For the standard3 SKU, the value is either \'default\' or \'highDensity\'. For all other SKUs, this value must be \'default\'.')
+@allowed([
+  'default'
+  'highDensity'
+])
+param hostingMode string = 'default'
```
Comment (ahmadabdalla) [inline modules/search/search-service/main.bicep:65]: Can we test the `hostingMode` on `highDensity`? Maybe the common tests include this?

File: modules/search/search-service/main.bicep [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+])
+param lock string = ''
+
+@description('Optional. Network specific rules that determine how the Azure Cognitive Search service may be reached.')
+param networkRuleSet object = {}
+
+@description('Optional. The number of partitions in the search service; if specified, it can be 1, 2, 3, 4, 6, or 12. Values greater than 1 are only valid for standard SKUs. For \'standard3\' services with hostingMode set to \'highDensity\', the allowed values are between 1 and 3.')
+@minValue(1)
+@maxValue(12)
+param partitionCount int = 1
```
Comment (ahmadabdalla) [inline modules/search/search-service/main.bicep:84]: Can we test with higher partition count? Maybe in the common file

File: modules/search/search-service/main.bicep [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+@allowed([
+  'enabled'
+  'disabled'
+])
+param publicNetworkAccess string = 'enabled'
+
+@description('Optional. The number of replicas in the search service. If specified, it must be a value between 1 and 12 inclusive for standard SKUs or between 1 and 3 inclusive for basic SKU.')
+@minValue(1)
+@maxValue(12)
+param replicaCount int = 1
```
Comment (ahmadabdalla) [inline modules/search/search-service/main.bicep:102]: Can we test with higher replica count? maybe in the common test file

File: modules/search/search-service/shared-private-link-resource/main.bicep [hunk 12]
```diff
@@ -0,0 +1,68 @@
+metadata name = 'Cognitive Search Private Link Resources'
```
Comment (ahmadabdalla) [inline modules/search/search-service/shared-private-link-resource/main.bicep:1]:
```suggestion
metadata name = 'Search Services Private Link Resources'
```

File: modules/search/search-service/shared-private-link-resource/main.bicep [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+metadata description = 'This module deploys a Cognitive Search Private Link Resource.'
```
Comment (ahmadabdalla) [inline modules/search/search-service/shared-private-link-resource/main.bicep:2]:
```suggestion
metadata description = 'This module deploys a Search Service Private Link Resource.'
```

**Review by ahmadabdalla — CHANGES_REQUESTED** [review]
Thanks @krbar ! Can you please sync the latest main before generating the final readmes.

File: modules/search/search-service/.test/common/main.test.bicep [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        {
+          value: '40.74.28.0/23'
+        }
+        {
+          value: '87.147.204.13'
+        }
+      ]
+    }
+    tags: {
+      Environment: 'Non-Prod'
```
Comment (AlexanderSehr) [inline modules/search/search-service/.test/common/main.test.bicep:106]:
```suggestion
      'hidden-title': 'This is visible in the resource name'
      Environment: 'Non-Prod'
```

File: modules/search/search-service/.test/pe/dependencies.bicep [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    addressSpace: {
+      addressPrefixes: [
+        addressPrefix
+      ]
+    }
+    subnets: [
+      {
+        name: 'defaultSubnet'
+        properties: {
+          addressPrefix: addressPrefix
```
Comment (AlexanderSehr) [inline modules/search/search-service/.test/pe/dependencies.bicep:34]:
```suggestion
          addressPrefix: cidrSubnet(addressPrefix, 16, 0)
```

File: modules/search/search-service/.test/pe/main.test.bicep [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        requestMessage: 'Please approve this request'
+      }
+      {
+        privateLinkResourceId: nestedDependencies.outputs.keyVaultResourceId
+        groupId: 'vault'
+        requestMessage: 'Please approve this request'
+      }
+    ]
+    tags: {
+      Environment: 'Non-Prod'
```
Comment (AlexanderSehr) [inline modules/search/search-service/.test/pe/main.test.bicep:91]:
```suggestion
      'hidden-title': 'This is visible in the resource name'
      Environment: 'Non-Prod'
```

File: modules/search/search-service/main.bicep [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+    requestMessage: contains(sharedPrivateLinkResource, 'requestMessage') ? sharedPrivateLinkResource.requestMessage : ''
+    resourceRegion: contains(sharedPrivateLinkResource, 'resourceRegion') ? sharedPrivateLinkResource.resourceRegion : ''
+  }
+}]
+
+// =========== //
+//   Outputs   //
+// =========== //
+
+@description('The name of the searchService.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/main.bicep:264]:
```suggestion
@description('The name of the search service.')
```

File: modules/search/search-service/main.bicep [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+output name string = searchService.name
+
+@description('The resource ID of the searchService.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/main.bicep:267]:
```suggestion
@description('The resource ID of the search service.')
```

File: modules/search/search-service/main.bicep [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+output resourceId string = searchService.id
+
+@description('The name of the resource group the searchService was created in.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/main.bicep:270]:
```suggestion
@description('The name of the resource group the search service was created in.')
```

File: modules/search/search-service/main.bicep [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+@batchSize(1)
+module searchService_sharedPrivateLinkResources 'shared-private-link-resource/main.bicep' = [for (sharedPrivateLinkResource, index) in sharedPrivateLinkResources: {
+  name: '${uniqueString(deployment().name, location)}-searchService-SharedPrivateLink-${index}'
+  params: {
+    name: contains(sharedPrivateLinkResource, 'name') ? sharedPrivateLinkResource.name : 'spl-${last(split(searchService.id, '/'))}-${sharedPrivateLinkResource.groupId}-${index}'
+    searchServiceName: searchService.name
+    privateLinkResourceId: sharedPrivateLinkResource.privateLinkResourceId
+    groupId: contains(sharedPrivateLinkResource, 'groupId') ? sharedPrivateLinkResource.groupId : ''
+    requestMessage: contains(sharedPrivateLinkResource, 'requestMessage') ? sharedPrivateLinkResource.requestMessage : ''
+    resourceRegion: contains(sharedPrivateLinkResource, 'resourceRegion') ? sharedPrivateLinkResource.resourceRegion : ''
```
Comment (AlexanderSehr) [inline modules/search/search-service/main.bicep:256]:
```suggestion
    resourceRegion: contains(sharedPrivateLinkResource, 'resourceRegion') ? sharedPrivateLinkResource.resourceRegion : ''
    enableDefaultTelemetry: enableReferencedModulesTelemetry
```

File: modules/search/search-service/shared-private-link-resource/main.bicep [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    requestMessage: !empty(requestMessage) ? requestMessage : null
+    resourceRegion: !empty(resourceRegion) ? resourceRegion : null
+  }
+}
+
+// =========== //
+//   Outputs   //
+// =========== //
+
+@description('The name of the sharedPrivateLinkResource.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/shared-private-link-resource/main.bicep:61]:
```suggestion
@description('The name of the shared private link resource.')
```

File: modules/search/search-service/shared-private-link-resource/main.bicep [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+output name string = sharedPrivateLinkResource.name
+
+@description('The resource ID of the sharedPrivateLinkResource.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/shared-private-link-resource/main.bicep:64]:
```suggestion
@description('The resource ID of the shared private link resource.')
```

File: modules/search/search-service/shared-private-link-resource/main.bicep [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+output resourceId string = sharedPrivateLinkResource.id
+
+@description('The name of the resource group the sharedPrivateLinkResource was created in.')
```
Comment (AlexanderSehr) [inline modules/search/search-service/shared-private-link-resource/main.bicep:67]:
```suggestion
@description('The name of the resource group the shared private link resource was created in.')
```

File: modules/search/search-service/main.bicep (hunk 7 above)
Comment (krbar) [reply modules/search/search-service/main.bicep]: Hi @ahmadabdalla, thanks for the quick response and review. The diagnostic params have been moved down as suggested.

File: modules/search/search-service/main.bicep (hunk 9 above)
Comment (krbar) [reply modules/search/search-service/main.bicep:34]: Added to common test, which was successful.

File: modules/search/search-service/main.bicep (hunk 10 above)
Comment (krbar) [reply modules/search/search-service/main.bicep:53]: Added to common test case as suggested, test was successful.

File: modules/search/search-service/main.bicep (hunk 11 above)
Comment (krbar) [reply modules/search/search-service/main.bicep:71]: Added to common test case as suggested, test was successful.

File: modules/search/search-service/main.bicep (hunk 8 above)
Comment (krbar) [reply modules/search/search-service/main.bicep:27]: Added to common test case as suggested, test was successful.

**Review by ahmadabdalla — APPROVED** [review]

**Pull request merged**